### ISSUE TITLE
GHC generates link-time warnings on Darwin with LLVM/clang 4.0.0

### ISSUE THREAD

**Issue by jwiegley** [NixOS/nixpkgs#25139]
@peti I'm noticing the following whenever I link anything using `cabal build` within a `nix-shell` environment. Note that I only started seeing this recently (though I don't know exactly when; I mistakenly thought it was due to a plugin I was using in another project). Does this have anything to do with the way the Nix environment is being set up?

```
Linking dist/build/solver/solver ...

In file included from /var/folders/ds/nt2q1_s57cqgt9g94_vmkjcw0000gn/T/ghc48015_0/ghc_1.c:1:0: error:


In file included from /nix/store/2wrxp3h0zwc1d50j67m8j792jknpgq84-ghc-8.0.2/lib/ghc-8.0.2/include/Rts.h:217:0: error:


/nix/store/2wrxp3h0zwc1d50j67m8j792jknpgq84-ghc-8.0.2/lib/ghc-8.0.2/include/rts/storage/ClosureMacros.h:505:5: error:
     warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
#if ZERO_SLOP_FOR_LDV_PROF || ZERO_SLOP_FOR_SANITY_CHECK
    ^

/nix/store/2wrxp3h0zwc1d50j67m8j792jknpgq84-ghc-8.0.2/lib/ghc-8.0.2/include/rts/storage/ClosureMacros.h:502:37: error:
     note: expanded from macro 'ZERO_SLOP_FOR_LDV_PROF'
#define ZERO_SLOP_FOR_LDV_PROF     (defined(PROFILING))
                                    ^

/nix/store/2wrxp3h0zwc1d50j67m8j792jknpgq84-ghc-8.0.2/lib/ghc-8.0.2/include/rts/storage/ClosureMacros.h:505:31: error:
     warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
#if ZERO_SLOP_FOR_LDV_PROF || ZERO_SLOP_FOR_SANITY_CHECK
                              ^

/nix/store/2wrxp3h0zwc1d50j67m8j792jknpgq84-ghc-8.0.2/lib/ghc-8.0.2/include/rts/storage/ClosureMacros.h:503:37: error:
     note: expanded from macro 'ZERO_SLOP_FOR_SANITY_CHECK'
#define ZERO_SLOP_FOR_SANITY_CHECK (defined(DEBUG) && !defined(THREADED_RTS))
                                    ^

/nix/store/2wrxp3h0zwc1d50j67m8j792jknpgq84-ghc-8.0.2/lib/ghc-8.0.2/include/rts/storage/ClosureMacros.h:505:31: error:
     warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]

/nix/store/2wrxp3h0zwc1d50j67m8j792jknpgq84-ghc-8.0.2/lib/ghc-8.0.2/include/rts/storage/ClosureMacros.h:503:56: error:
     note: expanded from macro 'ZERO_SLOP_FOR_SANITY_CHECK'
#define ZERO_SLOP_FOR_SANITY_CHECK (defined(DEBUG) && !defined(THREADED_RTS))
                                                       ^

/nix/store/2wrxp3h0zwc1d50j67m8j792jknpgq84-ghc-8.0.2/lib/ghc-8.0.2/include/rts/storage/ClosureMacros.h:523:5: error:
     warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
#if ZERO_SLOP_FOR_LDV_PROF && !ZERO_SLOP_FOR_SANITY_CHECK
    ^

/nix/store/2wrxp3h0zwc1d50j67m8j792jknpgq84-ghc-8.0.2/lib/ghc-8.0.2/include/rts/storage/ClosureMacros.h:502:37: error:
     note: expanded from macro 'ZERO_SLOP_FOR_LDV_PROF'
#define ZERO_SLOP_FOR_LDV_PROF     (defined(PROFILING))
                                    ^

/nix/store/2wrxp3h0zwc1d50j67m8j792jknpgq84-ghc-8.0.2/lib/ghc-8.0.2/include/rts/storage/ClosureMacros.h:523:32: error:
     warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
#if ZERO_SLOP_FOR_LDV_PROF && !ZERO_SLOP_FOR_SANITY_CHECK
                               ^

/nix/store/2wrxp3h0zwc1d50j67m8j792jknpgq84-ghc-8.0.2/lib/ghc-8.0.2/include/rts/storage/ClosureMacros.h:503:37: error:
     note: expanded from macro 'ZERO_SLOP_FOR_SANITY_CHECK'
#define ZERO_SLOP_FOR_SANITY_CHECK (defined(DEBUG) && !defined(THREADED_RTS))
                                    ^

/nix/store/2wrxp3h0zwc1d50j67m8j792jknpgq84-ghc-8.0.2/lib/ghc-8.0.2/include/rts/storage/ClosureMacros.h:523:32: error:
     warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]

/nix/store/2wrxp3h0zwc1d50j67m8j792jknpgq84-ghc-8.0.2/lib/ghc-8.0.2/include/rts/storage/ClosureMacros.h:503:56: error:
     note: expanded from macro 'ZERO_SLOP_FOR_SANITY_CHECK'
#define ZERO_SLOP_FOR_SANITY_CHECK (defined(DEBUG) && !defined(THREADED_RTS))
                                                       ^

/nix/store/2wrxp3h0zwc1d50j67m8j792jknpgq84-ghc-8.0.2/lib/ghc-8.0.2/include/rts/storage/ClosureMacros.h:552:5: error:
     warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
#if ZERO_SLOP_FOR_LDV_PROF && !ZERO_SLOP_FOR_SANITY_CHECK
    ^

/nix/store/2wrxp3h0zwc1d50j67m8j792jknpgq84-ghc-8.0.2/lib/ghc-8.0.2/include/rts/storage/ClosureMacros.h:502:37: error:
     note: expanded from macro 'ZERO_SLOP_FOR_LDV_PROF'
#define ZERO_SLOP_FOR_LDV_PROF     (defined(PROFILING))
                                    ^

/nix/store/2wrxp3h0zwc1d50j67m8j792jknpgq84-ghc-8.0.2/lib/ghc-8.0.2/include/rts/storage/ClosureMacros.h:552:32: error:
     warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
#if ZERO_SLOP_FOR_LDV_PROF && !ZERO_SLOP_FOR_SANITY_CHECK
                               ^

/nix/store/2wrxp3h0zwc1d50j67m8j792jknpgq84-ghc-8.0.2/lib/ghc-8.0.2/include/rts/storage/ClosureMacros.h:503:37: error:
     note: expanded from macro 'ZERO_SLOP_FOR_SANITY_CHECK'
#define ZERO_SLOP_FOR_SANITY_CHECK (defined(DEBUG) && !defined(THREADED_RTS))
                                    ^

/nix/store/2wrxp3h0zwc1d50j67m8j792jknpgq84-ghc-8.0.2/lib/ghc-8.0.2/include/rts/storage/ClosureMacros.h:552:32: error:
     warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]

/nix/store/2wrxp3h0zwc1d50j67m8j792jknpgq84-ghc-8.0.2/lib/ghc-8.0.2/include/rts/storage/ClosureMacros.h:503:56: error:
     note: expanded from macro 'ZERO_SLOP_FOR_SANITY_CHECK'
#define ZERO_SLOP_FOR_SANITY_CHECK (defined(DEBUG) && !defined(THREADED_RTS))
                                                       ^
9 warnings generated.
```

**Comment by henrytill** [comment]
Hi @jwiegley

I ran into this last week as well.  It appears to be related to the switch to LLVM/clang 4.0.0.

Looks like @shlevy identified this as a problem with GHC and fixed it upstream.  See https://phabricator.haskell.org/D2823

So it looks like the problem should be fixed in 8.2.1 and beyond, but will need to be addressed on our end for earlier versions...

**Comment by BilalReffas** [comment]
If you are using macOS then just uninstall the old ghc version and update ghc. You can use `uninstall-hs` for removing old versions.
